### DISCUSSION TITLE
How to test named delay on machine (without interpretation)

### DISCUSSION THREAD

**Discussion by FloNeu** [statelyai/xstate#4338] [Q&A] [answered]
I want to test a machine that has a named delay between two states like:
PENDING -> after:delay_befor_shown -> SHOWN
that i want to test with machine.transition() to make sure the transission 
exists and leads to SHOWN.

With and event transision, for example
HIDDEN -> event_show -> PENDING i could just do
machine.transition('HIDDEN', { type: 'event_show' }) 

how can i perform this structural test in a similar manner or do i have to change the approch
and use machine.getStateNode("PENDING") and then test if the delay exists on the state and
can't do it like with and external event?

I am aware that i could interprete the machine and use the simulated clock to test, but try to find
and understand my possebilities with the machine definitions only.

Regards, Florian

**Comment by davidkpiano** [comment, marked as answer]
Yes, interpreting the machine would be the best way to test this, but if you don't want to do that, remember that state machines are only affected by events, and that timeout is just an event!

You would have to construct the `"xstate.after..."` event, which is a bit internal. But here is the implementation in XState v5 beta, to give you an idea:

```ts
export function createAfterEvent(delayRef: number | string, id?: string) {
  const idSuffix = id ? `#${id}` : '';
  return { type: `xstate.after(${delayRef})${idSuffix}` } as const;
}
```

**Reply by FloNeu** [reply]
Very good - thank you very much... just testing out as much possebilties for a future xstate course/presentation. 

@_affected by events, and that timeout is just an event!_
 I thought so... but wasn't quite sure how to construct the type property... One more thing while i have you, regarding naming... when defined in the state machine diagram the lines between the states are called **transistions** no matter if its a event, a delay, always etc. - but the triggers for those transitions are then always correctly named **Events**, right?

And going by what you said i could also trigger an always transiton, by sending an event that has the right type-signature?

**Reply by davidkpiano** [reply]
That is correct!

However, an "always" transition is internal and not caused by an event.
